packages/react-client/src/ReactFlightClient.js

```diff
@@ -854,18 +900,90 @@ export function processBinaryChunk(
   response: Response,
   chunk: Uint8Array,
 ): void {
-  const stringDecoder = response._stringDecoder;
-  let linebreak = chunk.indexOf(10); // newline
-  while (linebreak > -1) {
-    const fullrow =
-      response._partialRow +
-      readFinalStringChunk(stringDecoder, chunk.subarray(0, linebreak));
-    processFullRow(response, fullrow);
-    response._partialRow = '';
-    chunk = chunk.subarray(linebreak + 1);
-    linebreak = chunk.indexOf(10); // newline
+  let i = 0;
+  while (i < chunk.length) {
+    let lastIdx = -1;
+    switch (response._rowState) {
+      case ROW_ID: {
+        const byte = chunk[i++];
+        if (byte === 58 /* ":" */) {
+          // Finished the rowID, next we'll parse the tag.
+          response._rowState = ROW_TAG;
+        } else {
+          response._rowID =
+            (response._rowID << 4) | (byte > 96 ? byte - 87 : byte - 48);
+        }
+        continue;
+      }
+      case ROW_TAG: {
+        const resolvedRowTag = chunk[i];
+        if (resolvedRowTag === 84 /* "T" */) {
+          response._rowTag = resolvedRowTag;
+          response._rowState = ROW_LENGTH;
+          i++;
+        } else if (resolvedRowTag > 64 && resolvedRowTag < 91 /* "A"-"Z" */) {
+          response._rowTag = resolvedRowTag;
+          response._rowState = ROW_CHUNK_BY_NEWLINE;
+          i++;
+        } else {
+          response._rowTag = 0;
+          response._rowState = ROW_CHUNK_BY_NEWLINE;
+          // This was an unknown tag so it was probably part of the data.
+        }
+        continue;
+      }
+      case ROW_LENGTH: {
+        const byte = chunk[i++];
+        if (byte === 44 /* "," */) {
+          // Finished the rowLength, next we'll buffer up to that length.
+          response._rowState = ROW_CHUNK_BY_LENGTH;
+        } else {
+          response._rowLength =
+            (response._rowLength << 4) | (byte > 96 ? byte - 87 : byte - 48);
+        }
+        continue;
+      }
+      case ROW_CHUNK_BY_NEWLINE: {
+        // We're looking for a newline
+        lastIdx = chunk.indexOf(10 /* "\n" */, i);
+        break;
+      }
+      case ROW_CHUNK_BY_LENGTH: {
+        // We're looking for the remaining byte length
+        const rowLength = response._rowLength;
+        if (i + rowLength <= chunk.length) {
+          lastIdx = i + rowLength;
+        }
+        break;
+      }
+    }
+    if (lastIdx > -1) {
+      // We found the last chunk of the row
+      const lastChunk = chunk.slice(i, lastIdx);
```

Comment:
Would DataView be more efficient here?

Reply:
Doubt it. It generally isn’t but also we don’t really control the source and this will also be used for binary streams where the protocol is uint8arrays too.

Reply:
Oh you mean to avoid the copy? I actually don’t know why I did a copy.

In general we expect these to be immutable representations so I could probably do a new array from the same underlying buffer.

It’ll need to be copied for serialized typed arrays in some cases anyway.

Reply:
FWIW, in JSC (don't know about V8), calling .subarray or .buffer on an ArrayBufferView will actually clone it on the first access because by default the lifetime of the buffer is the lifetime of the ArrayBufferView ("WastefulTypedArray" is the name for this in the code iirc)

Reply:
That's unfortunate. Given that TextDecoder requires a view, is there any workaround that doesn't?